DiffUtil & Repository

june0122 opened the pull request:
### 현재 작업 중인 내용 : DiffUtil 적용

### 문제점

- 어댑터 클래스 내부 `add`와 `remove` 메서드 내에서 `submitList`를 사용할 시 제대로 작동하지 않는 문제

```kotlin
/** UserListAdapter */

    override fun add(item: UserData) {
        newList.add(item)
//        submitList(newList) // submitList 사용 시 제대로 동작하지 않음
        this.notifyItemInserted(this.itemCount)
    }

    override fun remove(position: Int) {
        newList.removeAt(position)
//        submitList(newList) // submitList 사용 시 제대로 동작하지 않음
        this.notifyItemRemoved(position)
    }
```

june0122 committed:
refactor: DiffUtil 관련 클래스와 메서드 정의

june0122 committed:
refactor: paging 구현 관련 난해한 코드 수정

뷰모델 내부에 추가로 선언했던 itemCount 변수를 제거하고, 로딩 상태를 명확히 표시,
프로그래스바의 위치를 변수에 저장하는 시점과 프로그래스바로 스크롤하는 시점을 난해하지
않게끔 수정

june0122 committed:
refactor: 불필요한 옵저빙 삭제

june0122 committed:
refactor: DiffUtil 관련 코드 수정

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/utils/UserDiffCallback.kt:
```diff
@@ -0,0 +1,16 @@
+    }
+
+    override fun areContentsTheSame(oldItem: UserData, newItem: UserData): Boolean {
+        return oldItem == newItem
```
areItemsTheSame과 areContentsTheSame이 어떻게 다를까요?
실제 동작이 어떻게 다른지도 확인해보세요.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/adapter/UserListAdapter.kt:
```diff
@@ -39,40 +40,40 @@ class UserListAdapter(private val listener: UserClickListener) : RecyclerView.Ad
-        _items.add(item)
-        this.notifyItemInserted(_items.count())
+        newList.add(item)
+//        submitList(newList) // submitList 사용 시 제대로 동작하지 않음
```
왜 정상 동작 하지 않을까요?
https://thdev.tech/kotlin/2020/09/22/kotlin_effective_03/

june0122 committed:
refactor: bookmarks를 Event Wrapper로 감싸기

viewLifeCycle로 인해 두 번 동작하는 bookmarks를 Event Wrapper를 사용하여 해결
(SingleLiveEvent)

june0122 committed:
style: 리스트 및 상세 화면 레이아웃 수정

june0122 committed:
refactor: ListAdapter를 사용하지 않고 DiffUtil 적용

june0122 committed:
refactor: 리스트 관련 메서드들을 정의한 인터페이스 제거

DiffUtil을 적용하여 리스트 관련 메서드들을 정의한 인터페이스 제거

june0122 committed:
refactor: 리스트를 어댑터 내부가 아닌 뷰모델에서 LiveData로 관리하도록 변경

기존에 다른 동작이나 데이터를 LiveData를 옵저빙하는 방식으로 처리하고 있었는데
리스트의 데이터를 처리하는 방식도 LiveData를 사용하는 방식으로 변경하여
데이터를 처리하는 방식에 일관성을 줌.

june0122 committed:
style: 리사이클러뷰 아이템의 북마크 버튼을 위한 리소스 추가

june0122 committed:
refactor: ViewModel에 정의된 UI 노출 관련 코드 제거

ViewModel에 정의된 Snackbar 및 ImageResource와 같은 UI 노출 관련 코드를
View에서 노출하도록 수정.

ViewModel은 판단만 하고, 화면에 노출하는 작업은 View에서 이루어져야 한다.

june0122 committed:
feat: 리스트의 아이템에 북마크 버튼 기능 추가 (진행 중)

화면에 다시 돌아와도 북마크가 지정된 아이템을 표시하도록 수정해야 함.

june0122 committed:
fix: 북마크 화면의 포지션이 아닌 뷰모델 내부의 리스트 포지션이 들어가는 문제 수정

june0122 committed:
refactor: 패키지 구조 및 이름 리팩토링

june0122 committed:
chore: room 관련 라이브러리 추가

june0122 committed:
chore: Application 클래스 추가로 인한 매니페스트 업데이트

june0122 committed:
refactor: 패키지 위치를 network로 변경

june0122 committed:
refactor: 아키텍처 구성요소(Room, Repository) 추가

june0122 committed:
refactor: repository를 ViewModel의 매개변수로 추가

june0122 committed:
refactor: Room을 사용하여 entity를 통해 테이블 생성

june0122 committed:
refactor: 로컬 데이터베이스를 통해 북마크 기능 구현

june0122 committed:
refactor: 불필요한 옵저빙을 통한 북마크 구현 코드들 삭제
